### ISSUE TITLE
Outdated resources

### ISSUE THREAD

**Issue by kvaps** [LINBIT/linstor-server#106]
Hi, time to time we're facing with the next situation:

```
# linstor r l -r pvc-897fbae7-4230-4725-a97f-14d51734653b
+-----------------------------------------------------------------------------+
| ResourceName                             | Node  | Port | Usage  |    State |
|=============================================================================|
| pvc-897fbae7-4230-4725-a97f-14d51734653b | m15c1 | 8819 | Unused | Outdated |
| pvc-897fbae7-4230-4725-a97f-14d51734653b | m15c2 | 8819 | Unused | UpToDate |
+-----------------------------------------------------------------------------+
```

the problem is that linstor doesn't know anything about the third primary resource:
```bash
m15c1:~# drbdadm status pvc-897fbae7-4230-4725-a97f-14d51734653b
pvc-897fbae7-4230-4725-a97f-14d51734653b role:Secondary
  disk:Outdated
  m15c2 role:Secondary
    peer-disk:UpToDate

```

```bash
m15c2:~# drbdadm status pvc-897fbae7-4230-4725-a97f-14d51734653b
pvc-897fbae7-4230-4725-a97f-14d51734653b role:Secondary
  disk:UpToDate
  m10c1 role:Primary
    peer-disk:Diskless
  m15c1 role:Secondary
    peer-disk:Outdated
```

```bash
# drbdadm status pvc-897fbae7-4230-4725-a97f-14d51734653b
pvc-897fbae7-4230-4725-a97f-14d51734653b role:Primary
  disk:Diskless
  m15c1 connection:Connecting
  m15c2 role:Secondary
    peer-disk:UpToDate
```

I don't know how and why it was happened, but when I try to fix this situation by defining existing resource in linstor, I getting the error:

```
# linstor r c m10c1 pvc-897fbae7-4230-4725-a97f-14d51734653b --diskless
ERROR:
Description:
    (Node: 'm10c1') Shutdown of the DRBD resource 'pvc-897fbae7-4230-4725-a97f-14d51734653b failed
Cause:
    The external command for stopping the DRBD resource failed
Correction:
    - Check whether the required software is installed
    - Check whether the application's search path includes the location
      of the external software
    - Check whether the application has execute permission for the external command
Show reports:
    linstor error-reports show 5E3BE836-3A43A-000028
ERROR:
    (Node: 'm15c2') Failed to adjust DRBD resource pvc-897fbae7-4230-4725-a97f-14d51734653b
Show reports:
    linstor error-reports show 5E3BE839-9A434-000028
SUCCESS:
    Update of resource 'pvc-897fbae7-4230-4725-a97f-14d51734653b' on m10c1 applied on node 'm15c1'
```

[5E3BE836-3A43A-000028.log](https://github.com/LINBIT/linstor-server/files/4173199/5E3BE836-3A43A-000028.log)

And resource switching to the DELETING state:
```bash
# linstor r l -r pvc-897fbae7-4230-4725-a97f-14d51734653b
+-----------------------------------------------------------------------------+
| ResourceName                             | Node  | Port | Usage  |    State |
|=============================================================================|
| pvc-897fbae7-4230-4725-a97f-14d51734653b | m10c1 | 8819 |        | DELETING |
| pvc-897fbae7-4230-4725-a97f-14d51734653b | m15c1 | 8819 | Unused | Outdated |
| pvc-897fbae7-4230-4725-a97f-14d51734653b | m15c2 | 8819 | Unused | UpToDate |
+-----------------------------------------------------------------------------+
```

**Comment by ghernadi** [comment]
Right now there is only one scenario I can think of creating the behavior you are showing:

* create `rsc` (diskful) on node `A`
* create `rsc` (diskful) on node `B`
  * linstor will automagically create a tie-breaker `rsc` (diskless) on node `C`. That tie-breaker will not be visible by default in `r l` (you have to use `-a` or `--all` to see them). 
* something mounts the tie-breaker on node `C`, making it primary (auto-promote)
* for whatever reason something (or someone :) ) tries to manually remove the tie-breaker with `linstor r d C rsc`
  * the first node (in this case `A`) succeeds with removing its peer node `C`
  * DRBD on the second node `B` refuses to delete its peer `C` as `C` is diskless and `B` is the only remaining diskful.

this explains the scenario of the first part of your issue.

Then you are trying to manually create the diskless `rsc` on `C` (which internally still exists and is  marked as a tie-breaker resource, but already with the `DELETE` flag). Manually creating a resource "over" a tie-breaker simply removes the tie-breaker flag, effectively converting the same resource from linstor-managed to user-managed.   
However, that will not affect the `DELETE` flag which is still set in this scenario, which explains the errors of your `r c ..` command as well as the resulting (visible) diskless resource in `DELETING` state.

I don't really want to go into details, but this hafl-broken setup also explains why the DRBD-resource on one diskful node is `Outdated` while the other is `UpToDate`. 

Linstor now really wants to delete that diskless resource, and right now you cannot undo a deletion in linstor. That means, in order to resolve this issue, you have to allow linstor to remove that diskless resource (i.e. make the DRBD resource secondary and let linstor do its job). If you afterwards re-create the diskless resource, all 3 DRBD-resources should have a connection to each other and everything should be fine.

**Comment by kvaps** [comment]
Yes, I was thinking about this case, thanks for detailed explanation!
Looks like it is linstor-csi driver issue, because I see exact volumeattachmets for those nodes in our kubernetes cluster.

I can tell more, it seems that kubernetes csi-driver uses the rest-api interface, but the command for listing the resources and command for listing **all** the resources, is pretty same in API:

```
# linstor --curl r l 
curl -X GET http://localhost:3370/v1/view/resources
# linstor --curl r l -a
curl -X GET http://localhost:3370/v1/view/resources
```

**Comment by ghernadi** [comment]
You are right. That is the very same API, no hidden / subtle differences.

The filtering regarding the tie-breaker only happens in the client - i.e. when rendering the table. That means, if you look at the machine-readable output (with `-m`), you also should see the `TIE_BREAKER` flag on the resources. The `-a` flag only affects the filtering on the linstor-client site.

**Comment by kvaps** [comment]
So is it linstor-csi driver (and other clients) or linstor-server issue?

**Comment by ghernadi** [comment]
Either I am missing some information or I simply dont see right now how this should be a linstor-server issue. I will ping our maintainer of the linstor-csi driver to take a look on this issue.

**Comment by kvaps** [comment]
All drivers implements the next logic:
Eg. you need to crate resource `res1` on node `node1`

1. Check if `node1` is having resource `res1` created
2. If don't create `res1` on `node1`

The problem is that mostly all of them using the same api call to list the resources, there is no server-side filtering, thus tiebreaker resources are showing the same way like any other.

That's why drivers trying to use them, because they are "thinking" that tiebreaker is normal resource, but actually it is don't, that's why I'm thinking filtering should be implemented on server side.

This is not only `linstor-csi` driver issue. My one [`linstor_un`](https://github.com/OpenNebula/addon-linstor_un) is also have it. I presume that other official drivers have no implemented this filtering too:
- https://github.com/OpenNebula/addon-linstor
- https://github.com/LINBIT/openstack-cinder
- https://github.com/LINBIT/linstor-proxmox

**Comment by ghernadi** [comment]
I dont think this is a big of a problem - if the driver chooses the already existing diskless resource (which might be a tie-breaker) - okay, why not. If the driver chooses to deploy its own diskless resource then linstor will remove the (now) unnecessary tie-breaker. 

I dont see issues with this.

The root cause of your issue is that you somehow managed to start a deletion of the tie-breaker while it was primary on DRBD. 
I just ran a simple test (manually) and the controller actually rejects the deletion if it knows that the specific satellite actively uses that resource (i.e. that drbd-resource is primary). 
Which brings us back to the core question - what happened that you ended up in a half-deleted tie-breaking resource?

Is this test somehow reproducable? even if there is a race-condition, can you briefly tell me what the test does?

**Comment by ghernadi** [comment]
small update: I have found a reproducible way triggering this bug in linstor (or more in linstor's concept of handling tie-breaker). 
Your suggestion of introducing something like
```
curl -X GET http://localhost:3370/v1/view/resources?all=true
```
makes much more sense to me now.

**Comment by kvaps** [comment]
Cool, because I wasn't able to reproduce that 🙂

But I found another problem connected with that issue, maybe you're talking about the same:

enable tiebreaker:
```bash
# linstor c sp DrbdOptions/auto-add-quorum-tiebreaker true
```

create persistentvolume claim
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test
  uid: d3ae3937-0e19-4a48-a28e-5cd3327c96ae
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: linstor-1
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 10Gi
```

find the resource name:
```bash
# kubectl get pvc test
NAME   STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
test   Bound    pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5   10Gi       RWO            linstor-1      35s
```

find tiebreaker node:
```bash
# linstor r l -a -r pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5    
+-------------------------------------------------------------------------------+
| ResourceName                             | Node  | Port | Usage  |      State |
|===============================================================================|
| pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 | m1c10 | 7357 | Unused | TieBreaker |
| pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 | m1c12 | 7357 | Unused |   UpToDate |
| pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 | m1c4  | 7357 | Unused |   UpToDate |
+-------------------------------------------------------------------------------+
```

Create a pod on tiebreaker node
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
  - command:
    - sleep
    - infinity
    image: docker.io/library/alpine:3.11
    name: test
    volumeMounts:
    - mountPath: /data
      name: data
  nodeName: m1c10
  volumes:
  - name: data
    persistentVolumeClaim:
      claimName: test
```


```bash
# linstor r l -a -r pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 
+-------------------------------------------------------------------------------+
| ResourceName                             | Node  | Port | Usage  |      State |
|===============================================================================|
| pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 | m1c10 | 7357 | InUse  | TieBreaker |
| pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 | m1c12 | 7357 | Unused |   UpToDate |
| pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 | m1c4  | 7357 | Unused |   UpToDate |
+-------------------------------------------------------------------------------+
```

remove the pod
```bash
# kubectl delete pod test
```

the tiebraker resource removed too:

```bash
# linstor r l -a -r pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 
+-------------------------------------------------------------------------------+
| ResourceName                             | Node  | Port | Usage  |      State |
|===============================================================================|
| pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 | m1c12 | 7357 | Unused |   UpToDate |
| pvc-7c52a9f3-c331-42ab-8b4f-4f6cc734eaa5 | m1c4  | 7357 | Unused |   UpToDate |
+-------------------------------------------------------------------------------+
```

**Comment by ghernadi** [comment]
This is what happens:

the csi driver finds the diskless resource (tie-breaker is only a "linstor-managed" diskless resource), so it does not tries to create a new one (creating a resource over an existing user-managed resource causes an error, and I guess that csi simply does not know about the tie-breaker stuff), but starts using it (this is the state of `InUse` and `TieBreaker`). 

Once k8s is finished with the resource, it removes the diskless resource as it already forgot (or never remembered) that this resource already existed and was NOT created by csi.
Deleting a tie-breaking resource in general also sets the property `DrbdOptions/auto-add-quorum-tiebreaker` to false to prevent an endless fight between user's delete action and linstors auto-creating-tiebreaker. 

However, we are thinking about a fix of the original issue that should also fix this "strange" behavior. With the intended fix you your last `r l -a` command would show 3 resources, but the third would show a state of (user-managed) `Diskless` instead of a (linstor-managed) `TieBreaker`.

**Comment by kvaps** [comment]
fixed in v1.4.3
